File: sdk/go/tipsyabconfig/health.go
package tipsyabconfig

import (
	"sync"
	"time"
)

// BackgroundErrorEvent describes a single background-link failure delivered to
// Config.OnBackgroundError. The callback is invoked synchronously on the SDK's
// background goroutine; see Config.OnBackgroundError for the contract.
type BackgroundErrorEvent struct {
	// Phase is the background link that failed; one of:
	//   "startup_pull"  — the startup PullAll sweep failed and was absorbed
	//                      by StartupFailOpen (aggregate; Namespace is "").
	//   "periodic_pull" — a periodic (10 s fallback) PullAll for Namespace failed.
	//   "subscribe"     — the Subscribe stream errored and will reconnect.
	Phase string
	// Namespace is the namespace associated with the failure. It is "" for the
	// aggregate startup_pull failure and for subscribe failures (the stream is
	// multi-namespace).
	Namespace string
	// Err is the underlying error.
	Err error
	// Time is when the SDK observed the failure.
	Time time.Time
}

// Health is a snapshot of the SDK's background-link health. Obtain a copy via
// Client.Health. The zero value means "all healthy / no failures observed".
//
// HTTP transport mode (Config.Transport == TransportHTTP): the Subscribe stream
// is never opened, so SubscribeConnected is always false and LastSubscribeErr
// is always nil; the SDK relies solely on periodic PullAll, so only LastPullErr
// / StartupCacheEmpty are meaningful there.
type Health struct {
	// StartupCacheEmpty is true when StartupFailOpen absorbed a startup PullAll
	// failure and the client started with an empty cache.
	StartupCacheEmpty bool
	// LastPullErr is the most recent periodic PullAll error (nil = no failure
	// recorded yet). It is not cleared on subsequent success.
	LastPullErr error
	// LastPullErrTime is when LastPullErr was recorded.
	LastPullErrTime time.Time
	// LastSubscribeErr is the most recent Subscribe stream error (nil = none).
	// Always nil in HTTP transport mode (no Subscribe stream).
	LastSubscribeErr error
	// LastSubscribeErrTime is when LastSubscribeErr was recorded.
	LastSubscribeErrTime time.Time
	// SubscribeConnected reflects whether the Subscribe stream is currently
	// established (true after a successful stream open, false after a stream
	// error until the next successful open). Always false in HTTP transport
	// mode (no Subscribe stream is opened).
	SubscribeConnected bool
}

// healthState is the mutable, mutex-protected backing store for Client.Health.
// It is owned by the Client and accessed from the pull loop, the subscribe
// loop, Init, and external Health calls; all access goes through the helpers
// below so it is data-race-free under `go test -race`.
type healthState struct {
	mu   sync.Mutex
	snap Health
}

func (h *healthState) snapshot() Health {
	h.mu.Lock()
	defer h.mu.Unlock()
	return h.snap
}

func (h *healthState) setStartupCacheEmpty() {
	h.mu.Lock()
	h.snap.StartupCacheEmpty = true
	h.mu.Unlock()
}

func (h *healthState) recordPullErr(err error, t time.Time) {
	h.mu.Lock()
	h.snap.LastPullErr = err
	h.snap.LastPullErrTime = t
	h.mu.Unlock()
}

func (h *healthState) recordSubscribeErr(err error, t time.Time) {
	h.mu.Lock()
	h.snap.LastSubscribeErr = err
	h.snap.LastSubscribeErrTime = t
	h.snap.SubscribeConnected = false
	h.mu.Unlock()
}

func (h *healthState) setSubscribeConnected(connected bool) {
	h.mu.Lock()
	h.snap.SubscribeConnected = connected
	h.mu.Unlock()
}

// Health returns a snapshot of the SDK's background-link health. Safe for
// concurrent use.
func (c *Client) Health() Health {
	return c.health.snapshot()
}

// fireBackgroundError records the failure into health state and invokes the
// user's OnBackgroundError callback synchronously. The callback is wrapped in
// recover() so that a panicking consumer callback cannot kill the SDK's
// background goroutine — an intentional boundary exception to the "trust
// internal code" norm, since this callback is external consumer code.
//
// The callback stays synchronous by design: a slow callback will slow the
// pull/subscribe loop, which is documented on Config.OnBackgroundError.
func (c *Client) fireBackgroundError(ev BackgroundErrorEvent) {
	switch ev.Phase {
	case "startup_pull":
		c.health.setStartupCacheEmpty()
	case "periodic_pull":
		c.health.recordPullErr(ev.Err, ev.Time)
	case "subscribe":
		c.health.recordSubscribeErr(ev.Err, ev.Time)
	}
	cb := c.cfg.OnBackgroundError
	if cb == nil {
		return
	}
	defer func() { _ = recover() }()
	cb(ev)
}

File: sdk/java/tipsy-abconfig/src/main/java/io/github/lightspeedintelligence/abconfig/Health.java
package io.github.lightspeedintelligence.abconfig;

import java.time.Instant;
import java.util.Optional;

/**
 * Immutable snapshot of the SDK's background-link health.
 *
 * <p>Mirrors Go {@code Health}. Obtain a copy via the client's {@code health()}.
 * The "all healthy" state means no failures observed:
 * {@link #startupCacheEmpty()} false, both {@code lastErr} empty with their
 * times at {@link Instant#EPOCH}.
 *
 * <p>Error stickiness (mirrors Go): once {@link #lastPullErr()} /
 * {@link #lastSubscribeErr()} is recorded it is NOT cleared on a subsequent
 * success; only {@link #subscribeConnected()} flips with the stream connecting
 * / disconnecting. When no error has been recorded, the corresponding
 * {@code *Time()} returns {@link Instant#EPOCH} (a fixed zero value, never
 * {@code null}).
 *
 * <p>HTTP transport mode: the Subscribe stream is never opened, so
 * {@link #subscribeConnected()} is always false and {@link #lastSubscribeErr()}
 * is always empty.
 */
public final class Health {

    private final boolean startupCacheEmpty;
    private final Throwable lastPullErr;
    private final Instant lastPullErrTime;
    private final Throwable lastSubscribeErr;
    private final Instant lastSubscribeErrTime;
    private final boolean subscribeConnected;

    Health(boolean startupCacheEmpty,
           Throwable lastPullErr,
           Instant lastPullErrTime,
           Throwable lastSubscribeErr,
           Instant lastSubscribeErrTime,
           boolean subscribeConnected) {
        this.startupCacheEmpty = startupCacheEmpty;
        this.lastPullErr = lastPullErr;
        this.lastPullErrTime = lastPullErrTime;
        this.lastSubscribeErr = lastSubscribeErr;
        this.lastSubscribeErrTime = lastSubscribeErrTime;
        this.subscribeConnected = subscribeConnected;
    }

    /** True when {@code startupFailOpen} absorbed a startup PullAll failure and the client started empty. */
    public boolean startupCacheEmpty() {
        return startupCacheEmpty;
    }

    /** The most recent periodic PullAll error, or empty if none recorded (sticky). */
    public Optional<Throwable> lastPullErr() {
        return Optional.ofNullable(lastPullErr);
    }

    /** When {@link #lastPullErr()} was recorded, or {@link Instant#EPOCH} if none. */
    public Instant lastPullErrTime() {
        return lastPullErrTime;
    }

    /** The most recent Subscribe stream error, or empty if none (always empty in HTTP mode; sticky). */
    public Optional<Throwable> lastSubscribeErr() {
        return Optional.ofNullable(lastSubscribeErr);
    }

    /** When {@link #lastSubscribeErr()} was recorded, or {@link Instant#EPOCH} if none. */
    public Instant lastSubscribeErrTime() {
        return lastSubscribeErrTime;
    }

    /** Whether the Subscribe stream is currently established (always false in HTTP mode). */
    public boolean subscribeConnected() {
        return subscribeConnected;
    }
}
